pypeit/edgetrace.py

```diff
@@ -5014,41 +5074,135 @@ def order_refine_free_format(self, reference_row, combined_order_tol=1.8, debug=
         # and exclude the overlap regions from the reduction.
-        if np.any(_left[1:] - _right[:-1] < 0):
-            # Loop sequentially so that each pair is updated as the loop progresses
-            for i in range(1, _left.size):
-                # *Negative* of the gap; i.e., positives values means there's
-                # overlap
-                ngap = _right[i-1] - _left[i]
-                if ngap > 0:
-                    # Adjust both order edges to avoid the overlap region but
-                    # keep the same center coordinate
-                    _left[i-1] += ngap
-                    _right[i-1] -= ngap
-                    _left[i] += ngap
-                    _right[i] -= ngap
+        if np.all(_left[1:] - _right[:-1] > 0):
+            # There is no overlap, so just return the orders to add
+            return add_left, add_right, rmtraces
+
+        # Used to remove orders that have too much overlap
+        nord = _left.size
+        good_order = np.ones(nord, dtype=bool)
+
+        # Loop sequentially so that each pair is updated as the loop progresses
+        for i in range(1, nord):
+            # *Negative* of the gap; i.e., positives values means there's
+            # overlap
+            ngap = _right[i-1] - _left[i]
+            if ngap > 0:
+                if self.par['max_overlap'] is not None:
+                    good_order[i-1] = 2*ngap/(_right[i-1] - _left[i-1]) < self.par['max_overlap']
+                    good_order[i] = 2*ngap/(_right[i] - _left[i]) < self.par['max_overlap']
+                # Adjust both order edges to avoid the overlap region but
+                # keep the same center coordinate
+                _left[i-1] += ngap
+                _right[i-1] -= ngap
+                _left[i] += ngap
+                _right[i] -= ngap
+
+        # For any *existing* traces that were adjusted because of the overlap,
+        # this applies the adjustment to the `edge_fit` data.
+        # NOTE: This only adjusts the "fit" locations (edge_fit), *not* the
+        # measured centroid locations (edge_cen).  This should not cause
+        # problems because, e.g., the `get_slits` function uses `edge_fit`.
+        left_indx = np.where(left_gpm)[0][individual_orders]
+        offset = _left[isrt][add_left.size:] - left
+        self.edge_fit[:,left_indx] += offset[None,:]
+        right_indx = np.where(right_gpm)[0][individual_orders]
+        offset = _right[isrt][add_left.size:] - right
+        self.edge_fit[:,right_indx] += offset[None,:]
 
         # Get the adjusted traces to add.  Note this currently does *not* change
         # the original traces
-        return _left[isrt][:add_left.size], _right[isrt][:add_right.size], rmtraces
-        
+        good_order = good_order[isrt][:add_left.size]
+        add_left = _left[isrt][:add_left.size][good_order]
+        add_right = _right[isrt][:add_left.size][good_order]
```

Comment:
Hey @kbwestfall it seems there is bug here (many ADAP datasets failed here). Should this line be:
`add_right = _right[isrt][:add_right.size][good_order]`

Reply:
Ah, yes.  Sorry about this!  I (hopefully) fixed this and pushed a new version.